synapse/config/server.py

```diff
@@ -666,6 +750,48 @@ def default_config(self, server_name, data_dir_path, **kwargs):
         # 'true'.
         #
         #show_users_in_user_directory: false
+
+        # Message retention policy at the server level.
+        #
+        # Room admins and mods can define a retention period for their rooms using the
+        # 'im.vector.room.retention' state event, and server admins can cap this period
+        # by setting the 'min_lifetime' and 'max_lifetime' config options. This feature is
+        # disabled by default.
+        #
+        # If this feature is enabled, Synapse will regularly look for and purge events
+        # which are older than the room's maximum retention period. Synapse will also
+        # filter events received over federation so that events that should have been
+        # purged are ignored and not stored again.
+        #
+        # Rooms that lack the 'im.vector.room.retention' state event will be considered as
+        # if their maximum retention period is equal to the 'max_lifetime' parameter from
+        # the server's configuration.
+        #
+        # Server admins can also define the settings of the background jobs purging the
+        # events which lifetime has expired under the 'purge_jobs' section.
+        # If no configuration is provided, a single job will be setup to delete expired
+        # events in every room daily.
+        # Each job's configuration defines which range of message lifetime the job takes
+        # care of. For example, if 'min_lifetime' is '2d' and 'max_lifetime' is '3d', the
+        # job will handle purging expired events in rooms which state defines a
+        # 'max_lifetime' that's between 2 and 3 days.
+        #
+        # The rationale for this per-job configuration is that server admins might want to
+        # run the purge job at a low interval for rooms with low 'max_lifetime', but want
+        # to increase the interval for rooms with higher 'max_lifetime' in order to ease
```

Comment:
in any case, I'd be inclined to have an initial version with a single job, and then consider adding multiple jobs later.

Reply:
I've had a chat with @erikjohnston about this, since he suggested that bit, and it turns out that I initially didn't understand correctly the rationale behind it.

Broadly, it's that we're likely to end up with rooms that have a retention period of small periods like one day which are expected to be purged fairly frequently (e.g. every 30min; the final value for this hasn't been decided yet), because it's fairly OK for a message having a lifetime of one year to live one year and a half, but not really for a message with a lifetime of one day to live for two days. And if we end up with tens of thousands of rooms to purge, we probably don't want to load retention data for all of the rooms the homeserver knows about every 5min/30min/1h/etc. because it'll end up being quite expensive.

Therefore, the idea is to have e.g. one job that handles rooms which history have a small lifetime, which will run fairly frequently and will only pull data on these rooms, and have another one handle all of the other rooms.

I don't think there was a hard requirement for this to be configurable, but I thought I'd make it so in order to offer more granularity and make it easier to manage in the future should DINSIC or someone else want to tune the frequencies and targets of these jobs.

Reply:
I'll fix the comment.